### PR TITLE
Implemented VK sync plugins

### PR DESCRIPTION
**Pull request by antseppa**

### CODE REVIEW

File: src/vk/vk-contacts/vkcontactsyncadaptor.cpp [hunk 1]
```diff
@@ -0,0 +1,1010 @@
...
+}
+
+QString VKContactSyncAdaptor::syncServiceName() const
+{
+    return QStringLiteral("vk-contacts");
+}
+
+void VKContactSyncAdaptor::sync(const QString &dataTypeString, int accountId)
+{
+    m_apiRequestsRemaining[accountId] = 99; // assume we can make up to 99 requests per sync, before being throttled.
```
Comment (chriadam) [inline src/vk/vk-contacts/vkcontactsyncadaptor.cpp]: I think this is incorrect.  I think the limits are MUCH smaller than this, although I cannot remember what the limits are at the moment.

File: src/vk/vk-contacts/vkcontactsyncadaptor.cpp [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+    } else if (data.isEmpty()) {
+        SOCIALD_LOG_ERROR("no contact data in reply from VK with account:" << accountId);
+        purgeSyncStateData(QString::number(accountId));
+        setStatus(SocialNetworkSyncAdaptor::Error);
+        decrementSemaphore(accountId);
+        return;
+    }
+
+    // parse the remote contact information from the response
+    QString continuationUrl;
```
Comment (chriadam) [inline src/vk/vk-contacts/vkcontactsyncadaptor.cpp]: this continuationUrl is never set, we never parse it.  I guess when I originally wrote this, I wasn't able to test with an account that had more than about 4 friends... We need to figure out what this should be, and parse it properly.

File: src/vk/vk-contacts/vkcontactsyncadaptor.cpp [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+    QJsonObject obj = QJsonDocument::fromJson(data).object(); // XXX TODO: check errors / continuations / etc.
```
Comment (chriadam) [inline src/vk/vk-contacts/vkcontactsyncadaptor.cpp]: Also, error checking here would be good too ;-)

File: src/vk/vk-notifications/vknotificationsyncadaptor.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+void VKNotificationSyncAdaptor::finalize(int accountId)
+{
+    Q_UNUSED(accountId);
+    if (syncAborted()) {
+        SOCIALD_LOG_DEBUG("sync aborted, skipping finalize of VK Notifications from account:" << accountId);
+    } else {
+        SOCIALD_LOG_DEBUG("finalizing VK Notifications sync with account:" << accountId);
+        //m_db.purgeOldNotifications(OLD_NOTIFICATION_LIMIT_IN_DAYS); // TODO
```
Comment (chriadam) [inline src/vk/vk-notifications/vknotificationsyncadaptor.cpp]: is this still TODO or has it been implemented some other way?

File: src/vk/vk-posts/vk-posts.pri [hunk 5]
```diff
@@ -0,0 +1,5 @@
+CONFIG += link_pkgconfig meegotouchevents-qt5
```
Comment (chriadam) [inline src/vk/vk-posts/vk-posts.pri]: is meegotouchevents-qt5 still a thing?

File: src/vk/vk-posts/vkpostsyncadaptor.cpp [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+            if (!object.isEmpty()) {
+                if (object.value(QStringLiteral("type")).toString() == QStringLiteral("post")) {
+                    m_postsToAdd.append(PostData(accountId, object, userProfiles, groupProfiles));
+                } else  if (object.value(QStringLiteral("type")).toString() == QStringLiteral("photo")) {
+                    m_photoPostsToAdd.append(PostData(accountId, object, userProfiles, groupProfiles));
+                } else if (object.value(QStringLiteral("type")).toString() == QStringLiteral("photo_tag")) {
+                    SOCIALD_LOG_DEBUG("TODO: unhandled newsfeed item type:" << object.value(QStringLiteral("type")).toString() << ", skipping.");
+                } else if (object.value(QStringLiteral("type")).toString() == QStringLiteral("wall_photo")) {
+                    SOCIALD_LOG_DEBUG("TODO: unhandled newsfeed item type:" << object.value(QStringLiteral("type")).toString() << ", skipping.");
+                } else if (object.value(QStringLiteral("type")).toString() == QStringLiteral("note")) {
```
Comment (chriadam) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: we should probably handle note and wall_photo post types if possible.  It shouldn't be too tricky, I hope?

File: src/vk/vk-posts/vkpostsyncadaptor.cpp [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+    }
+
+    // VK post indentifier is just index number and not globally unique. To make
+    // it unique we combine it with fromId
+    QString identifier = QString::number(fromId) + QStringLiteral("_") +
+                         (post.contains(QStringLiteral("id"))
+                            ? QString::number(post.value(QStringLiteral("id")).toDouble())
+                            : QString::number(post.value(QStringLiteral("post_id")).toDouble()));
+
+    newPost.link = QStringLiteral("https://m.vk.com/wall") + identifier;
```
Comment (chriadam) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: should that be https://m.vk.com/wall/ + identifier?  missing / separator?

File: src/vk/vk-posts/vkpostsyncadaptor.cpp [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+    }
+
+    // VK post indentifier is just index number and not globally unique. To make
+    // it unique we combine it with fromId
+    QString identifier = QString::number(fromId) + QStringLiteral("_") +
+                         (post.contains(QStringLiteral("id"))
+                            ? QString::number((int)post.value(QStringLiteral("id")).toDouble())
+                            : QString::number((int)post.value(QStringLiteral("post_id")).toDouble()));
+
+    newPost.link = QStringLiteral("https://m.vk.com/album")
```
Comment (chriadam) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: again, do we need / separator?  or is this correct?

File: src/vk/vk-posts/vkpostsyncadaptor.cpp [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+    }
+    MGConfItem heightConf("/lipstick/screen/primary/height");
+    if (heightConf.value(defaultValue).toInt() != defaultValue) {
+        height = heightConf.value(defaultValue).toInt();
+    }
+
+    // we want to use the largest of these dimensions as the "optimal"
+    int maxDimension = qMax(width, height);
+    if (maxDimension >= 2048) {
+        m_optimalImageSize = "photo_1280";
```
Comment (chriadam) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: why 1280 if dimension is 2048?

File: src/vk/vk-posts/vkpostsyncadaptor.cpp (hunk 7 above)
Comment (antseppa) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: no, the format really is wall123456789 without separator.

File: src/vk/vk-posts/vkpostsyncadaptor.cpp (hunk 8 above)
Comment (antseppa) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: same as above

File: src/vk/vk-posts/vkpostsyncadaptor.cpp (hunk 9 above)
Comment (antseppa) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: That was the highgest the spec mentioned, but there actually is higher one in the server reply.

File: src/vk/vk-posts/vkpostsyncadaptor.cpp (hunk 6 above)
Comment (antseppa) [inline src/vk/vk-posts/vkpostsyncadaptor.cpp]: true, it shouldn't, I just drew the line here to get initial version merged with support for most common cases.

File: src/vk/vk-posts/vk-posts.pri (hunk 5 above)
Comment (antseppa) [inline src/vk/vk-posts/vk-posts.pri]: a leftover, I'll remove it.

File: src/vk/vk-contacts/vkcontactsyncadaptor.cpp (hunk 2 above)
Comment (antseppa) [inline src/vk/vk-contacts/vkcontactsyncadaptor.cpp]: I have the same problem, need to test this with someone who has enough friends in VK.

File: src/vk/vk-contacts/vkcontactsyncadaptor.cpp (hunk 1 above)
Comment (antseppa) [inline src/vk/vk-contacts/vkcontactsyncadaptor.cpp]: The API documentation says two per second and 5000 per hour https://vk.com/pages?oid=-17680044&p=Limitations_for_Calling_Ads_API_Methods

**Comment by antseppa** [comment]
Updated:
- squashed to one commit
- adapted to use VKImageDatabase const ptr pattern
- fixed review comments

Still TODO:
- contacts sync adaptor needs some work:
     * filter out empty contacts
     * implement & test continuationUrl mechanism
     * make sure we won't exceed two requests-per-second limit with the above
- notifications sync adaptor purge

**Comment by chriadam** [comment]
Not sure that squashing down to one commit is a good idea, but I guess it doesn't matter.

**Comment by antseppa** [comment]
Had the original structure still left, decided to unsquash after discussion. Otherwise same.

**Comment by chriadam** [comment]
LGTM.  We can fix the issues with contact sync adaptor once we can test that one, I guess.

**Comment by antseppa** [comment]
Updated: rebased

**Comment by antseppa** [comment]
Updated: only sync those contacts which have contact details.

**Comment by antseppa** [comment]
Updated: rebased

**Comment by chriadam** [comment]
Still LGTM

**Pull request merged**